File: lib/src/nusheader.rs
use std::fmt::Display;

use crate::{buffer::Header, error::Error};

pub type Crc = (u32, u32);

#[derive(Debug)]
pub struct NusHeader {
    cfg_flags: u32,
    clck_rate: u32,
    boot_addr: u32,
    lu_ver: u32,
    crc: Crc,
    reserved_1: u64,
    // will be cut off after 0x14 bytes,
    // therefore spilling into the
    // reserved part after
    title: String,
    reserved_2: [u8; 0x07],
    category: u8,
    unique: [u8; 2],
    destination: u8,
    version: u8,
}

impl Display for NusHeader {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        writeln!(f, "cfg flags: {:x}", self.cfg_flags)?;
        writeln!(f, "clock rate: {:x}", self.clck_rate)?;
        writeln!(f, "lu version: {:x}", self.lu_ver)?;
        writeln!(f, "boot address: {:x}", self.boot_addr)?;
        writeln!(f, "crc: ({:x}-{:x})", self.crc.0, self.crc.1)?;
        writeln!(f, "title: {}", self.title)?;
        writeln!(f, "category: {}", self.category as char)?;
        writeln!(
            f,
            "category: {}{}",
            self.unique[0] as char, self.unique[1] as char
        )?;
        writeln!(f, "destination: {}", self.destination as char)?;
        writeln!(f, "version: {}", self.version)
    }
}

impl Default for NusHeader {
    fn default() -> Self {
        Self {
            cfg_flags: 0x80_37_12_40,
            clck_rate: 0xF,
            boot_addr: 0x80010000,
            lu_ver: Default::default(),
            crc: Default::default(),
            reserved_1: Default::default(),
            title: Default::default(),
            reserved_2: Default::default(),
            category: b'N',
            unique: Default::default(),
            destination: b'A',
            version: Default::default(),
        }
    }
}

const HEADER_SIZE: usize = 0x40;
const CRC_START: usize = 0x1000;
const CRC_LEN: usize = 0x100000;
const TITLE_LEN: usize = 0x14;

impl Header for NusHeader {
    /// Create a nus header from raw bytes
    fn from_bytes(data: &[u8]) -> Result<Self, Error>
    where
        Self: Sized,
    {
        if data.len() < HEADER_SIZE {
            return Err(Error::HeaderNotEnoughData);
        }

        // TODO this uses a lot of magic offsets
        // but it should be fine because this *literally* will never change (tm)
        Ok(Self {
            cfg_flags: Self::try_u32_from_be(data, 0x00)?,
            clck_rate: Self::try_u32_from_be(data, 0x04)?,
            boot_addr: Self::try_u32_from_be(data, 0x08)?,
            lu_ver: Self::try_u32_from_be(data, 0x0C)?,
            crc: (
                Self::try_u32_from_be(data, 0x10)?,
                Self::try_u32_from_be(data, 0x14)?,
            ),
            reserved_1: Self::try_u64_from_be(data, 0x18)?,
            // 0x14 bytes are reserved for the title
            // it is assumed that it is ascii only!
            title: String::from_utf8(data[0x20..0x34].to_vec())?,
            reserved_2: data[0x34..0x3B].try_into()?,
            category: data[0x3B],
            unique: data[0x3C..0x3E].try_into()?,
            destination: data[0x3E],
            version: data[0x3F],
        })
    }

    /// Converts the NusHeader to
    /// a byte array
    fn to_bytes(&self) -> Vec<u8> {
        let mut header = vec![];

        header.extend(self.cfg_flags.to_be_bytes());
        header.extend(self.clck_rate.to_be_bytes());
        header.extend(self.boot_addr.to_be_bytes());
        header.extend(self.lu_ver.to_be_bytes());
        header.extend(self.crc.0.to_be_bytes());
        header.extend(self.crc.1.to_be_bytes());
        header.extend(self.reserved_1.to_be_bytes());

        // Title always needs to be exactly 0x14 bytes
        let title = format!("{:0<20}", self.title);
        header.extend(&title.as_bytes()[0..TITLE_LEN]);

        header.extend(self.reserved_2);
        header.push(self.category);
        header.extend(self.unique);
        header.push(self.destination);
        header.push(self.version);

        // this panic is justified because
        // it should never occur that the header
        // length is not the correct size, and
        // if it is for some reason it absolutely is a bug
        if header.len() != HEADER_SIZE {
            panic!(
                "Expected vec of length {} but got {}!",
                HEADER_SIZE,
                header.len()
            );
        }

        header
    }

    /// Same as default, but calculates the crc
    fn default_with_crc(&mut self, data: &[u8]) -> Result<Self, Error>
    where
        Self: Sized,
    {
        Ok(Self {
            crc: Self::crc(data)?,
            ..Default::default()
        })
    }

    /// Calculate the nus crc and set the value in the header
    fn set_crc(&mut self, data: &[u8]) -> Result<(), Error> {
        let crc = Self::crc(data)?;
        self.crc = crc;
        Ok(())
    }
}

impl NusHeader {
    fn try_u32_from_be(data: &[u8], at: usize) -> Result<u32, Error> {
        Ok(u32::from_be_bytes(data[at..at + 4].try_into()?))
    }

    fn try_u64_from_be(data: &[u8], at: usize) -> Result<u64, Error> {
        Ok(u64::from_be_bytes(data[at..at + 8].try_into()?))
    }

    /// Calculates the nus crc sum
    /// similar to libdragon's implementation of chksm64
    // TODO clean this up and make it as readable as possible
    pub fn crc(data: &[u8]) -> Result<Crc, Error> {
        const INITIAL: u32 = 0x3f_u32.wrapping_mul(0x5d588b65) + 1;
        const T5: u32 = 32;

        if data.len() < CRC_LEN + CRC_START {
            return Err(Error::CrcNotEnoughData);
        }

        // required outside of loop
        let mut t8;
        let mut t6;
        let mut a3 = INITIAL;
        let mut s0 = INITIAL;

        let mut t7;
        let mut t9;

        let mut t2 = INITIAL;
        let mut t3 = INITIAL;
        let mut a2 = INITIAL;
        let mut t4 = INITIAL;

        for t1 in (0..CRC_LEN).step_by(4) {
            let current_data = u32::from_be_bytes(
                data[CRC_START + t1 as usize..CRC_START + t1 as usize + 4].try_into()?,
            );
            let v1 = a3.wrapping_add(current_data);
            let a1 = v1;

            if v1 < a3 {
                t2 = t2.wrapping_add(1);
            }

            let v1 = current_data & 0x1F;
            t7 = T5.wrapping_sub(v1);
            t8 = current_data >> t7;
            t6 = current_data << v1;
            let a0 = t6 | t8;
            a3 = a1;
            t3 ^= current_data;
            s0 = s0.wrapping_add(a0);

            if a2 < current_data {
                t9 = a3 ^ current_data;
                a2 ^= t9;
            } else {
                a2 ^= a0;
            }

            t7 = current_data ^ s0;
            t4 = t4.wrapping_add(t7);
        }

        // crc1
        t6 = a3 ^ t2;
        a3 = t6 ^ t3;

        // crc2
        t8 = s0 ^ a2;
        s0 = t8 ^ t4;

        Ok((a3, s0))
    }
}

#[cfg(test)]
mod test {
    use super::*;

    const CRC_END: usize = CRC_START + CRC_LEN;

    #[test]
    fn crc_fail() {
        assert_eq!(Err(Error::CrcNotEnoughData), NusHeader::crc(&[0, 1, 2]));

        // test failure off by 1
        let mut test_data = vec![];
        for i in 0..CRC_END - 1 {
            test_data.push(i as u8);
        }
        assert_eq!(Err(Error::CrcNotEnoughData), NusHeader::crc(&test_data));
    }

    #[test]
    fn crc() {
        let mut test_data = vec![];

        // 0, 1, 2, 3, ... 255, 0, 1, ...
        for i in 0..CRC_END {
            test_data.push(i as u8);
        }

        // result dervice from calculating the checksum
        // using the legacy python script
        assert_eq!(Ok((4207429594, 3000934689)), NusHeader::crc(&test_data));
    }
}

File: src/nusheader.c
#include "nusheader.h"
#include <string.h>
#include <stdlib.h>
#include <arpa/inet.h>

void nus_add_header(Buffer *buffer) {
  usize old_len = buffer->len;
  u8 *old_data = buffer->data;

  buffer->len += NUS_HEADER_SIZE;
  buffer->data = malloc(buffer->len);

  // this is different from the regular buffer resize function
  memset(buffer->data, 0, NUS_HEADER_SIZE);
  memcpy(buffer->data + NUS_HEADER_SIZE, old_data, old_len);

  free(old_data);
}

void nus_set_header(Buffer *buffer, NusHeader *header) {
  nus_crc(header, buffer->data, buffer->len);
  u8 header_bytes[NUS_HEADER_SIZE];
  nus_to_bytes(header, header_bytes);
  memcpy(buffer->data, header_bytes, NUS_HEADER_SIZE);
}

void nus_fprint(FILE *file, const NusHeader *header) {
  fprintf(file, "cfg flags: %x\n", header->cfg_flags);
  fprintf(file, "clock rate: %x\n", header->clck_rate);
  fprintf(file, "lu version: %x\n", header->lu_ver);
  fprintf(file, "boot address: %x\n", header->boot_addr);
  fprintf(file, "crc: (%x - %x)\n", header->crc.crc1, header->crc.crc2);

  fprintf(file, "title: ");
  // because the title may not be \0 terminated we need to
  // make sure to avoid an overrun
  for (usize i = 0; i < NUS_TITLE_LEN && header->title[i] != '\0'; i++) {
    fprintf(file, "%c", header->title[i]);
  }
  fprintf(file, "\n");

  fprintf(file, "category: %c\n", header->category);
  fprintf(file, "uid: %c%c\n", header->unique[0], header->unique[1]);
  fprintf(file, "destination: %c\n", header->destination);
  fprintf(file, "version: %d\n", header->version);
}

void nus_init(NusHeader *header) {
  memset(header, 0, sizeof(NusHeader));
  header->cfg_flags = 0x80371240; // NOLINT
  header->clck_rate = 0xF;        // NOLINT
  header->boot_addr = 0x80010000; // NOLINT
  header->category = 'N';
  header->destination = 'A';
}

u32 ntohl_from_(const u8 *data, usize offset) {
  return ntohl(*((u32 *)(data + offset)));
}

Error nus_from_bytes(NusHeader *header, const u8 *data, const usize len) {
  if (len < NUS_HEADER_SIZE) {
    return ERR_HEADER_NOT_ENOUGH_DATA;
  }

  // default values first!
  nus_init(header);

  // read the corresponding bytes
  header->cfg_flags = ntohl_from_(data, 0x00);
  header->clck_rate = ntohl_from_(data, 0x04);
  header->boot_addr = ntohl_from_(data, 0x08);
  header->lu_ver = ntohl_from_(data, 0x0C);
  header->crc.crc1 = ntohl_from_(data, 0x10);
  header->crc.crc2 = ntohl_from_(data, 0x14);

  memcpy(header->reserved_1, data + 0x18, 0x08);
  memcpy(header->title, data + 0x20, NUS_TITLE_LEN);
  memcpy(header->reserved_2, data + 0x34, 0x07);

  header->category = (char)data[0x3B];

  header->unique[0] = (char)data[0x3C];
  header->unique[1] = (char)data[0x3D];

  header->destination = (char)data[0x3E];
  header->version = (char)data[0x3F];

  // attempt crc, if it fails just leave it be!
  // if it didn't failt it should correct a wrong crc value
  // or calculate the same result as the exisiting value
  nus_crc(header, data, len);

  return OK;
}

// writes a BE u32 to result at offset
void htonl_to_(u32 n, u8 *result, const usize offset) {
  n = htonl(n);
  memcpy(result + offset, &n, sizeof(u32));
}

void nus_to_bytes(NusHeader *header, u8 *result) {
  htonl_to_(header->cfg_flags, result, 0x00);
  htonl_to_(header->clck_rate, result, 0x04);
  htonl_to_(header->boot_addr, result, 0x08);
  htonl_to_(header->lu_ver, result, 0x0C);
  htonl_to_(header->crc.crc1, result, 0x10);
  htonl_to_(header->crc.crc2, result, 0x14);

  memcpy(result + 0x18, header->reserved_1, 0x08);
  memcpy(result + 0x20, header->title, NUS_TITLE_LEN);
  memcpy(result + 0x34, header->reserved_2, 0x07);

  result[0x3B] = header->category;

  result[0x3C] = header->unique[0];
  result[0x3D] = header->unique[1];

  result[0x3E] = header->destination;
  result[0x3F] = header->version;
}

Error calc_crc_(const u8 *data, const usize len, NusCrc *crc) {
  // this is just some magic number used as an initial value
  const u32 INITIAL = -120959524;

  if (len < NUS_CRC_LEN + NUS_CRC_START) {
    return ERR_CRC_NOT_ENOUGH_DATA;
  }

  u32 t2 = INITIAL;
  u32 t3 = INITIAL;
  u32 a2 = INITIAL;
  u32 t4 = INITIAL;
  u32 crc1 = INITIAL;
  u32 crc2 = INITIAL;

  for (u32 idx = 0; idx < NUS_CRC_LEN; idx += 4) {
    // take the next 4 bytes and convert them to
    // a native integer
    u32 current_data = ntohl_from_(data, NUS_CRC_START + idx);

    u32 a1 = crc1 + current_data;

    if (a1 < crc1) {
      t2 = t2 + 1;
    }

    u32 v1 = current_data & 0x1f;                                // NOLINT
    u32 a0 = (current_data << v1) | (current_data >> (32 - v1)); // NOLINT
    crc1 = a1;

    t3 ^= current_data;

    crc2 = crc2 + a0;

    if (a2 < current_data) {
      a2 ^= crc1 ^ current_data;
    } else {
      a2 ^= a0;
    }

    t4 = t4 + (current_data ^ crc2);
  }

  crc1 = (crc1 ^ t2) ^ t3;
  crc2 = (crc2 ^ a2) ^ t4;

  crc->crc1 = crc1;
  crc->crc2 = crc2;

  return OK;
}

Error nus_crc(NusHeader *header, const u8 *data, const usize len) {
  NusCrc result;
  Error err = calc_crc_(data, len, &result);

  header->crc = result;

  return err;
}

#ifdef TEST

#include "macros.h"

void test_crc_fail(void **state) {
  u8 data1[] = {0, 1, 2, 3};
  NusCrc result;
  assert_int_equal(ERR_CRC_NOT_ENOUGH_DATA, calc_crc_(data1, 3, &result));

  // test failure off by 1 erro
  assert_int_equal(ERR_CRC_NOT_ENOUGH_DATA,
                   calc_crc_(data1, NUS_CRC_END - 1, &result));
}

void test_crc(void **state) {
  u8 *test_data = malloc(NUS_CRC_END);
  NusCrc result;

  for (u32 i = 0; i < NUS_CRC_END; i++) {
    test_data[i] = (u8)i;
  }

  assert_int_equal(OK, calc_crc_(test_data, NUS_CRC_END, &result));

  assert_int_equal(4207429594, result.crc1);
  assert_int_equal(3000934689, result.crc2);

  free(test_data);
}

#endif
